### PR TITLE
logic separation

### PR DESCRIPTION
**Pull request by urgobalt**
- feat: big brain logging
- feat: separation of functionality
- fix: log file location

### CODE REVIEW

File: parser/src/logging.rs
Comment (barbariand) [inline parser/src/logging.rs:1]: This should be from utils package right?

File: parser/Cargo.toml [hunk 1]
```diff
@@ -9,16 +9,14 @@ edition = "2021"
...
-clap = { version = "4.5.4", features = ["derive"] }
+clap = { version = "4.5.4", features = ["derive", "cargo"] }
 tracing = { version = "0.1.40", features = ["async-await"] }
 tracing-subscriber = { version = "0.3.18", features = ["env-filter"] }
+tracing-appender = {version = "0.2.3"}
 rustyline = "14.0.0"
 futures = "0.3.30"
 slicedisplay = "0.2.2"
 snafu = { version = "0.8.2", features = ["backtrace", "futures"] }
+directories = "5.0.1"
```
Comment (barbariand) [inline parser/Cargo.toml:15]: Add utils here so logging does not exist twies
```suggestion
utils = {path = "../utils"}
```

File: parser/src/lib.rs [hunk 2]
```diff
@@ -9,6 +9,7 @@ pub mod context;
 mod lexer;
 
 pub mod error;
+mod logging;
```
Comment (barbariand) [inline parser/src/lib.rs]: remove logging from parser and move it to utils
```suggestion

```

File: repl/src/main.rs [hunk 3]
```diff
@@ -1,45 +1,29 @@
...
-pub mod error;
-use prelude::*;
-mod normalizer;
-mod parsing;
-mod value;
-
-mod token;
-mod token_reader;
-use clap::{command, Parser as ClapParser};
+use clap::{command, crate_authors, crate_name, Parser as ClapParser};
```
Comment (barbariand) [inline repl/src/main.rs:3]: maybe dont need to use it as ClapParser but its also nice, just a thought

File: parser/src/logging.rs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            guard = Some(inner_guard);
+        }
+    }
+
+    tracing_subscriber::registry()
+        .with(stdout.and_then(log_file_writer))
+        .init();
+
+    guard
+}
```
Comment (barbariand) [inline parser/src/logging.rs]: delete this
```suggestion

```

File: parser/src/prelude.rs [hunk 5]
```diff
@@ -9,6 +9,7 @@ pub use crate::{
     approximator::Approximator,
     ast::Ast,
     context::{MathContext, MathFunction},
+    error::{AstError, ParseError},
```
Comment (barbariand) [inline parser/src/prelude.rs:12]: yea it should be exposed

File: parser/Cargo.toml [hunk 6]
```diff
@@ -9,16 +9,14 @@ edition = "2021"
 tokio = { version = "1.37.0", features = ["test-util", "tracing", "rt-multi-thread", "macros"] }
 async-recursion = "1.1.0"
 colored = "2.1.0"
```
Comment (barbariand) [inline parser/Cargo.toml]: this is not used
```suggestion
```

File: parser/Cargo.toml [hunk 7]
```diff
@@ -9,16 +9,14 @@ edition = "2021"
...
-clap = { version = "4.5.4", features = ["derive"] }
+clap = { version = "4.5.4", features = ["derive", "cargo"] }
 tracing = { version = "0.1.40", features = ["async-await"] }
 tracing-subscriber = { version = "0.3.18", features = ["env-filter"] }
+tracing-appender = {version = "0.2.3"}
 rustyline = "14.0.0"
 futures = "0.3.30"
 slicedisplay = "0.2.2"
 snafu = { version = "0.8.2", features = ["backtrace", "futures"] }
+directories = "5.0.1"
```
Comment (barbariand) [inline parser/Cargo.toml]: i dont think these are used here
```suggestion
futures = "0.3.30"
slicedisplay = "0.2.2"
snafu = { version = "0.8.2", features = ["backtrace", "futures"] }
```

File: utils/src/logging.rs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+};
+
+pub trait LogsDirectory {
+    fn log_dir(&self) -> Option<PathBuf>;
+}
+
+impl LogsDirectory for ProjectDirs {
+    fn log_dir(&self) -> Option<PathBuf> {
+        if cfg!(target_os = "linux") {
+            self.state_dir().map(|path| path.join("logs"))
```
Comment (barbariand) [inline utils/src/logging.rs:17]: why is this not the same as datadir as it is defined for linux (and macOS)

**Review by barbariand — CHANGES_REQUESTED** [review]
mostly removing dependancies that should not be needed anymore from my understanding.

File: parser/src/logging.rs
Comment (urgobalt) [reply parser/src/logging.rs:1]: I deleted it on local, dont know why it is pushed

Comment (barbariand) [reply parser/src/logging.rs:1]: Things hapen just push it again or smth

File: utils/src/logging.rs (hunk 8 above)
Comment (urgobalt) [reply utils/src/logging.rs:17]: There is amfeature request regarding this. I am simply using their definition to keep to standardizations

**Comment by urgobalt** [comment]
Should be fixed now.

**Review by barbariand — APPROVED** [review]
LGTM

**Comment by urgobalt** [comment]
Great, then I merge this!

**Pull request merged**